File: elixir/perfect_numbers.ex
defmodule PerfectNumbers do
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: {:ok, atom} | {:error, String.t()}
  def classify(number) when number < 1 do
    {:error, "Classification is only possible for natural numbers."}
  end

  def classify(number) do
    for i <- 1..(number - 1)//1, rem(number, i) == 0 do
      i
    end
    |> Enum.sum()
    |> then(fn n ->
      cond do
        n == number -> {:ok, :perfect}
        n < number -> {:ok, :deficient}
        true -> {:ok, :abundant}
      end
    end)
  end
end

File: zig/perfect_numbers.zig
pub const Classification = enum {
    deficient,
    perfect,
    abundant,
};

/// Asserts that `n` is nonzero.
pub fn classify(n: u64) Classification {
    var sum: u64 = 0;
    for (1..n) |i| {
        if (n % i == 0) {
            sum += i;
        }
    }

    if (sum < n) {
        return .deficient;
    }
    if (sum == n) {
        return .perfect;
    }
    return .abundant;
}
